File: camstudy_app/lib/viewmodels/study_room_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../models/chat_message.dart';
import '../models/study_room.dart';
import '../services/firestore_service.dart';

class StudyRoomViewModel extends ChangeNotifier {
  StudyRoom room;
  String? errorMessage;
  int elapsedSeconds = 0;
  List<ChatMessage> messages = [];
  String draftMessage = '';

  StreamSubscription<StudyRoom>? _roomSubscription;
  StreamSubscription<List<ChatMessage>>? _messagesSubscription;
  Timer? _timer;
  DateTime? _enteredAt;
  final _firestoreService = FirestoreService.shared;

  StudyRoomViewModel({required this.room});

  Future<void> enter({required String userId, required String nickname}) async {
    _enteredAt = DateTime.now();
    elapsedSeconds = 0;
    _startTimer();

    try {
      await _firestoreService.joinStudyRoom(
        roomId: room.id,
        userId: userId,
        nickname: nickname,
      );
    } catch (e) {
      errorMessage = e.toString();
      notifyListeners();
    }

    _observeRoom();
    _observeMessages();
  }

  Future<void> sendMessage({
    required String senderId,
    required String senderNickname,
  }) async {
    final trimmed = draftMessage.trim();
    if (trimmed.isEmpty) return;
    draftMessage = '';
    notifyListeners();

    try {
      await _firestoreService.sendMessage(
        roomId: room.id,
        senderId: senderId,
        senderNickname: senderNickname,
        text: trimmed,
      );
    } catch (e) {
      errorMessage = e.toString();
      notifyListeners();
    }
  }

  Future<void> leave(String userId) async {
    _stopTimer();
    await _saveSession(userId);

    try {
      await _firestoreService.leaveStudyRoom(roomId: room.id, userId: userId);
    } catch (e) {
      errorMessage = e.toString();
    }
  }

  void stopObserving() {
    _roomSubscription?.cancel();
    _roomSubscription = null;
    _messagesSubscription?.cancel();
    _messagesSubscription = null;
  }

  void _startTimer() {
    _timer = Timer.periodic(const Duration(seconds: 1), (_) {
      elapsedSeconds += 1;
      notifyListeners();
    });
  }

  void _stopTimer() {
    _timer?.cancel();
    _timer = null;
  }

  Future<void> _saveSession(String userId) async {
    final enteredAt = _enteredAt;
    if (enteredAt == null || elapsedSeconds <= 0) return;

    try {
      await _firestoreService.saveStudySession(
        userId: userId,
        roomId: room.id,
        roomName: room.name,
        startedAt: enteredAt,
        endedAt: DateTime.now(),
        durationSeconds: elapsedSeconds,
      );
      await _firestoreService.incrementTotalStudySeconds(
        userId: userId,
        seconds: elapsedSeconds,
      );
    } catch (e) {
      errorMessage = e.toString();
      notifyListeners();
    }
  }

  void _observeRoom() {
    _roomSubscription = _firestoreService
        .observeStudyRoom(room.id)
        .listen(
          (updatedRoom) {
            room = updatedRoom;
            notifyListeners();
          },
          onError: (Object error) {
            errorMessage = error.toString();
            notifyListeners();
          },
        );
  }

  void _observeMessages() {
    _messagesSubscription = _firestoreService
        .observeMessages(room.id)
        .listen(
          (updatedMessages) {
            messages = updatedMessages;
            notifyListeners();
          },
          onError: (Object error) {
            errorMessage = error.toString();
            notifyListeners();
          },
        );
  }

  @override
  void dispose() {
    _stopTimer();
    stopObserving();
    super.dispose();
  }
}

File: CamStudy/CamStudy/ViewModels/StudyRoomViewModel.swift
import Combine
import FirebaseFirestore
import Foundation

@MainActor
final class StudyRoomViewModel: ObservableObject {
    @Published var room: StudyRoom
    @Published var errorMessage: String?
    @Published var elapsedSeconds = 0
    @Published var messages: [ChatMessage] = []
    @Published var draftMessage = ""

    private var listener: ListenerRegistration?
    private var messagesListener: ListenerRegistration?
    private var timerCancellable: AnyCancellable?
    private var enteredAt: Date?
    private let firestoreService = FirestoreService.shared

    init(room: StudyRoom) {
        self.room = room
    }

    func enter(userId: String, nickname: String) {
        enteredAt = Date()
        elapsedSeconds = 0
        startTimer()

        firestoreService.joinStudyRoom(roomId: room.id, userId: userId, nickname: nickname) { [weak self] result in
            Task { @MainActor in
                if case .failure(let error) = result {
                    self?.errorMessage = error.localizedDescription
                }
            }
        }
        observeRoom()
        observeMessages()
    }

    func sendMessage(senderId: String, senderNickname: String) {
        let trimmed = draftMessage.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        let message = ChatMessage(
            id: UUID().uuidString,
            senderId: senderId,
            senderNickname: senderNickname,
            text: trimmed,
            sentAt: Date()
        )
        draftMessage = ""

        firestoreService.sendMessage(roomId: room.id, message: message) { [weak self] result in
            Task { @MainActor in
                if case .failure(let error) = result {
                    self?.errorMessage = error.localizedDescription
                }
            }
        }
    }

    func leave(userId: String) {
        stopTimer()
        saveSession(userId: userId)

        firestoreService.leaveStudyRoom(roomId: room.id, userId: userId) { [weak self] result in
            Task { @MainActor in
                if case .failure(let error) = result {
                    self?.errorMessage = error.localizedDescription
                }
            }
        }
    }

    func stopObserving() {
        listener?.remove()
        listener = nil
        messagesListener?.remove()
        messagesListener = nil
    }

    private func startTimer() {
        timerCancellable = Timer.publish(every: 1, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                self?.elapsedSeconds += 1
            }
    }

    private func stopTimer() {
        timerCancellable?.cancel()
        timerCancellable = nil
    }

    private func saveSession(userId: String) {
        guard let enteredAt, elapsedSeconds > 0 else { return }

        let session = StudySession(
            id: UUID().uuidString,
            userId: userId,
            roomId: room.id,
            roomName: room.name,
            startedAt: enteredAt,
            endedAt: Date(),
            durationSeconds: elapsedSeconds
        )

        firestoreService.saveStudySession(session) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success:
                    self.firestoreService.incrementTotalStudySeconds(userId: userId, by: session.durationSeconds) { result in
                        if case .failure(let error) = result {
                            Task { @MainActor in
                                self.errorMessage = error.localizedDescription
                            }
                        }
                    }
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    private func observeRoom() {
        listener = firestoreService.observeStudyRoom(roomId: room.id) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success(let room):
                    self.room = room
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    private func observeMessages() {
        messagesListener = firestoreService.observeMessages(roomId: room.id) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success(let messages):
                    self.messages = messages
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }
}
